File: add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode l3 = new ListNode();
        ListNode pointer = l3;
        boolean carry = false;
        while (l1 != null || l2 != null || carry) {
            int sum = 0;
            if (l1 != null && l2 != null) {
                sum = l1.val + l2.val + (carry ? 1 : 0);
            } else if (l1 != null) {
                sum = l1.val + (carry ? 1 : 0);
            } else if (l2 != null){
                sum = l2.val + (carry ? 1 : 0);
            } else {
                sum = carry ? 1 : 0;
            }
            if (sum >= 10) {
                sum -= 10;
                carry = true;
            } else {
                carry = false;
            }
            pointer.next = new ListNode(sum);
            pointer = pointer.next;
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        return l3.next;
    }
}

File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        carry = 0
        head = ListNode()
        pointer = head
        while l1 != None and l2 != None:
            sum = l1.val + l2.val + carry
            if sum > 9:
                sum -= 10
                carry = 1
            else:
                carry = 0
            node = ListNode(val = sum)
            pointer.next = node
            pointer = pointer.next
            l1 = l1.next
            l2 = l2.next
        while l1 != None:
            sum = l1.val + carry
            if sum > 9:
                sum -= 10
                carry = 1
            else:
                carry = 0
            node = ListNode(val = sum)
            pointer.next = node
            pointer = pointer.next
            l1 = l1.next
        while l2 != None:
            sum = l2.val + carry
            if sum > 9:
                sum -= 10
                carry = 1
            else:
                carry = 0
            node = ListNode(val = sum)
            pointer.next = node
            pointer = pointer.next
            l2 = l2.next
        if carry == 1:
            node = ListNode(val = 1)
            pointer.next = node
        return head.next

            
        